### PR TITLE
SyntaxArena: thread safety

### PR DESCRIPTION
**Pull request by rintaro**
Use mutex to guard concurrent mutation to the arena.
But make a "single thread" mode so that we don't regress performance too much when we know the arena is not used by other threads.

(non `Darwin`/`Glibc` environment e.g. Windows implementation is currently a dummy)

https://github.com/apple/swift-syntax/issues/785
rdar://99812330

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
@swift-ci Please test

**Comment by jpsim** [comment]
Thank you, this certainly helps, although TSan now reports an access race in the guard code:

```

WARNING: ThreadSanitizer: Swift access race (pid=42514)
  Modifying access of Swift variable at 0x0001148c22b0 by thread T25:
    #0 ScopeGuard.withGuard<A>(body:) <null>:82026728 (SwiftLintFrameworkTests:arm64+0x126a334)
    #1 SyntaxArena.withGuard<A>(body:) <null>:82026728 (SwiftLintFrameworkTests:arm64+0x126b504)
    #2 SyntaxArena.allocateRawSyntaxBuffer(count:) <null>:82026728 (SwiftLintFrameworkTests:arm64+0x126c2fc)
    #3 static RawSyntax.makeLayout(kind:uninitializedCount:arena:initializingWith:) <null>:82026728 (SwiftLintFrameworkTests:arm64+0xffcd14)
    #4 RawSyntaxLayoutView.replacingChild(at:with:arena:) <null>:82026728 (SwiftLintFrameworkTests:arm64+0x10036c8)
    #5 SyntaxData.replacingChild(_:at:) <null>:82026728 (SwiftLintFrameworkTests:arm64+0x1285864)
```

Full report: https://buildkite.com/swiftlint/swiftlint/builds/3411#018354b5-95eb-4c14-ab21-f7d59113f241

File: Sources/SwiftSyntax/SyntaxArena.swift [hunk 1]
```diff
@@ -114,7 +189,9 @@ public class SyntaxArena {
   /// Copies a `RawSyntaxData` to the memory this arena manages, and retuns the
   /// pointer to the destination.
   func intern(_ value: RawSyntaxData) -> UnsafePointer<RawSyntaxData> {
-    let allocated = allocator.allocate(RawSyntaxData.self, count: 1).baseAddress!
+    let allocated = lock.withGuard {
```
Comment (ahoppen) [inline Sources/SwiftSyntax/SyntaxArena.swift:192]: Same here, should this be `self.withGuard`?

File: Sources/SwiftSyntax/SyntaxArena.swift [hunk 2]
```diff
@@ -10,11 +10,60 @@
...
+/// Dummy mutex that doesn't actually guard at all.
+class ScopeGuard {
+  init() {}
+  func withGuard<T>(body: () throws -> T) rethrows -> T {
+    return try body()
+  }
+}
+#endif
+
 public class SyntaxArena {
```
Comment (CodaFi) [inline Sources/SwiftSyntax/SyntaxArena.swift:58]: Can we allocate these locks inline with the arena blocks?

**Review by CodaFi — CHANGES_REQUESTED** [review]
These ScopeGuards are not correct. We cannot use the materialization operator here despite the fact that we're indexing through a heap allocation for the member access - it's still a logical data race even if it's not always a physical data race. These locks need to be allocated out-of-line and the resulting address passed.

**Review by CodaFi — APPROVED** [review]

**Comment by rintaro** [comment]
@jpsim I confirmed the failure locally. And the modified implementation resolves it. Could you check the new implementation?

**Comment by rintaro** [comment]
@swift-ci Please test

**Review by jpsim — APPROVED** [review]
There are no more TSan failures in my testing. Thank you!

**Pull request merged**